### PR TITLE
#2521 Grid widget layout

### PR DESCRIPTION
**Pull request by AlexeyTuboltsev**
fix misaligned dashboard widgets and refactor the general widget layout to use css grid

### CODE REVIEW

File: app/components/shared/charts/ChartCircle.module.scss [hunk 1]
```diff
@@ -1,12 +1,12 @@
 .chart-circle {
   position: relative;
+  width: 100%;
+  max-width: 25rem;
```
Comment (desfero) [inline app/components/shared/charts/ChartCircle.module.scss]: px-to-rem maybe?

File: app/components/shared/charts/ChartCircle.module.scss [hunk 2]
```diff
@@ -1,12 +1,12 @@
 .chart-circle {
   position: relative;
+  width: 100%;
```
Comment (desfero) [inline app/components/shared/charts/ChartCircle.module.scss]: Do we need `width: 100%`?

File: app/components/shared/charts/ChartCircle.module.scss [hunk 3]
```diff
@@ -65,3 +64,13 @@
   max-width: 64%;
   text-align: center;
 }
+
+@media (min-width: 1500px) {
```
Comment (desfero) [inline app/components/shared/charts/ChartCircle.module.scss]: please use bootstrap variables for sizes, move @media to `.chart-circle` above and remove `position:relative` and `width:100%`

File: app/components/shared/Layout.module.scss [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+.span-2,
+.span-3,
+.span-4 {
+  grid-column: 1 / span 1;
+}
+
+
+@media (min-width: 770px) {
```
Comment (desfero) [inline app/components/shared/Layout.module.scss]: can we use bootstrap variables in media for consistency? Same in @media below

File: app/components/shared/charts/ChartDoughnut.module.scss [hunk 5]
```diff
@@ -15,6 +15,7 @@
   .chart-wrapper {
     width: 36%;
     margin: 0 auto;
+    max-width: 15rem;
```
Comment (desfero) [inline app/components/shared/charts/ChartDoughnut.module.scss]: `px-to-rem`?

File: app/components/eto/dashboard/UploadInvestmentMemorandum.tsx [hunk 6]
```diff
@@ -1,19 +1,30 @@
...
+    title={formatIntlMessage("settings.upload-investment-memorandum.title")}
+    text={<FormattedMessage id="settings.upload-investment-memorandum-please-upload-prospectus" />}
+    to={appRoutes.documents}
+    buttonText={<FormattedMessage id="settings.upload-investment-button.title" />}
+    layoutClass={layoutClass}
+  />
+);
+
+export const UploadInvestmentMemorandum = compose<React.FunctionComponent<IExternalProps>>(
+  createErrorBoundary(ErrorBoundaryPanel),
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/dashboard/UploadInvestmentMemorandum.tsx]: yes, this one doesn't need an error boundary. In general I think it's ok to leave error boundaries in widgets that have their own functionality, like in settings there are email widget, kyc widget, mnemonics widget, some of them make API calls that can go wrong etc, so if something breaks in one of them, like a failed http call, it;s good for UX to be able to use the rest of the app.

File: app/components/shared/PeopleSwiperWidget.unsafe.tsx [hunk 7]
```diff
@@ -92,13 +93,15 @@ class PeopleSwiperWidgeLayout extends React.PureComponent<IOwnProps & IDispatchP
...
-        visibleWidth: visibleWidth,
-      });
+    const [visibleWidth, elementWidth] = this.getVisibleWidth();
+    const newState = {
+      isVisible: !!this.swiperRef.current,
+      elementWidth,
+      visibleWidth,
+    };
+
+    if (isEqual(newState, this.state)) {
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/PeopleSwiperWidget.unsafe.tsx]: 😱

File: app/components/shared/charts/ChartCircle.module.scss (hunk 2 above)
Comment (AlexeyTuboltsev) [reply app/components/shared/charts/ChartCircle.module.scss]: yes, i want them to be big :))

File: app/components/eto/EtoDashboard.tsx [hunk 8]
```diff
@@ -38,6 +40,9 @@ import { UploadTermSheetWidget } from "./dashboard/UploadTermSheetWidget";
 import { DashboardSection } from "./shared/DashboardSection";
 import { EProjecStatusLayout, EProjectStatusSize, ETOState } from "./shared/ETOState";
 
+import * as layoutStyles from "../shared/Layout.module.scss";
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/EtoDashboard.tsx]: changed to an Enum

File: app/components/layouts/Layout.tsx [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+import * as styles from "./Layout.module.scss";
+
+export enum EColumnSpan {
+  ONE_COL = 1,
+  ONE_AND_HALF_COL,
+  TWO_COL,
+  THREE_COL,
+}
+
+export const resolveColumnSpan = (span: EColumnSpan = EColumnSpan.ONE_COL) => {
```
Comment (desfero) [inline app/components/layouts/Layout.tsx]: can we move it to `Container`?

File: app/components/layouts/Layout.module.scss [hunk 10]
```diff
@@ -0,0 +1,43 @@
+.widget-layout {
+  display: grid;
```
Comment (desfero) [inline app/components/layouts/Layout.module.scss]: IMO, `.widget-layout`s should not be a grid container, it should just be a child of `grid`.

File: app/components/layouts/Layout.tsx [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+      return styles.span4;
+    case EColumnSpan.THREE_COL:
+      return styles.span6;
+    case EColumnSpan.ONE_COL:
+    default:
+      return styles.span2;
+  }
+};
+
+export const WidgetGridLayout: React.FunctionComponent<{
```
Comment (desfero) [inline app/components/layouts/Layout.tsx]: can just make naming more consistent? I see in articles that grid child is often called `item` and parent `container`. We already call parent `Container` so maybe we can rename this to `Item`?

File: app/components/eto/EtoDashboard.tsx [hunk 12]
```diff
@@ -267,16 +283,16 @@ const EtoDashboard = compose<React.FunctionComponent>(
     }),
   }),
   withProps<IComputedProps, IStateProps>(props => ({
-    isVerificationSectionDone: !!(
+    isVerificationSectionDone: Boolean(
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/EtoDashboard.tsx]: no semantic reason, but with !! IntelliJ says this expression can be simplified etc, so it's just the way to get rid of red underline :)

File: app/components/eto/dashboard/ETOFormsProgressSection.module.scss [hunk 13]
```diff
@@ -1,17 +1,5 @@
...
-  .progress-section-layout {
-    grid-gap: 2rem;
-  }
-}
-
-@media (min-width: 990px) {
+@include media-breakpoint-between(sm, lg) {
   .progress-section-layout {
-    grid-template: auto / repeat(3, 1fr);
+    grid-template: auto / repeat(2, 1fr);
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/dashboard/ETOFormsProgressSection.module.scss]: but this is a one-off situation. This particular element (eto forms progress) should behave slightly different. I would move everything to one Enum on grid component, but it's for later. For now there's too big a mess, every page uses own layout etc. Let's do it later :))

File: app/components/shared/Panel.tsx [hunk 14]
```diff
@@ -20,10 +23,14 @@ const Panel: React.FunctionComponent<IPanelProps & CommonHtmlProps & TDataTestId
   className,
   children,
   narrow,
+  columnSpan,
   "data-test-id": dataTestId,
 }) => (
-  <section
-    className={cn(styles.panel, className, { [styles.narrow]: narrow })}
+  <Container
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/Panel.tsx]: Screen Shot 2019-04-18 at 14 42 48
Screen Shot 2019-04-18 at 14 42 58

File: app/components/eto/EtoDashboard.tsx [hunk 15]
```diff
@@ -214,20 +228,22 @@ const EtoDashboardComponent: React.FunctionComponent<
   } = props;
 
   return (
-    <GridBaseLayout dataTestId="eto-dashboard-application">
+    <WidgetGridLayout
+      data-test-id="eto-dashboard-application"
+      className={layoutStyles.layoutOffset}
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/EtoDashboard.tsx]: done as local .module.scss style

File: app/components/layouts/Layout.module.scss (hunk 10 above)
Comment (AlexeyTuboltsev) [reply app/components/layouts/Layout.module.scss]: done

File: app/components/wallet/Wallet.tsx [hunk 16]
```diff
@@ -2,14 +2,19 @@ import * as React from "react";
...
 import { createErrorBoundary } from "../shared/errorBoundary/ErrorBoundary.unsafe";
 import { ErrorBoundaryLayoutAuthorized } from "../shared/errorBoundary/ErrorBoundaryLayoutAuthorized";
 import { WalletRouter } from "./Router";
 
+import * as layoutStyles from "../layouts/Layout.module.scss";
+
 export const WalletComponent: React.FunctionComponent = () => (
   <LayoutAuthorized>
-    <WalletRouter />
+    <WidgetGridLayout data-test-id="wallet-start-container" className={layoutStyles.layoutOffset}>
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/wallet/Wallet.tsx]: done as prop

File: app/components/eto/dashboard/ETOFormsProgressSection.tsx [hunk 17]
```diff
@@ -146,20 +157,29 @@ export const ETOFormsProgressSectionComponent: React.FunctionComponent<IStatePro
...
-          dataTestId={section.testingId}
-          key={index}
-        />
+    <>
+      {groups.map((group, groupIndex) => (
+        <Container key={groupIndex} columnSpan={EColumnSpan.THREE_COL}>
+          <Heading level={2} decorator={false} className="mb-3">
+            {group.name}
+          </Heading>
+          <Container type={EContainerType.INHERIT_GRID} className={styles.progressSectionLayout}>
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/dashboard/ETOFormsProgressSection.tsx]: done

File: app/components/eto/dashboard/ETOFormsProgressSection.module.scss (hunk 13 above)
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/dashboard/ETOFormsProgressSection.module.scss]: the right way to do it is to have a one layout component with a couple of Enums to indicate settings, and all css stuff should be handled under the hood without devs touching css at all. But this is for later

**Comment by AlexeyTuboltsev** [comment]
Visual regressionchecked, all fine now

**Pull request merged**